pkg/cloud/client.go

```diff
@@ -141,8 +148,17 @@ func NewClientFromConf(conf Config) (Client, error) {
 	cacheMutex.Lock()
```

Comment:
do we still need this mutex given the threadsafe nature of the cache?

Reply:
Yes.  Otherwise it'd still be subject to the same concurrency problem described before, where multiple goroutines calling NewClientFromConf concurrently both discover the client they want isn't in the cache and then create it / add it.  The critical section here extends from the cache check to the cache put.

Reply:
@maxdrib, elaborating a little further, in your comment on the previous PR for this you asked whether this really mattered if both goroutines would put the same client to the cache (map).  Josh pointed out that the operation of putting the into the non-concurrency-safe map was itself risky (as it is a complex operation involving separate lookups and writes).  But why, then, wouldn't use of this concurrency-safe cache component not solve that?  It does address that particular challenge, but there's more.

The problem is that we really don't know that both clients are the same.   Several "constructor" calls are being made into libraries.  We don't really know if they're thread-safe, or if repeated calls truly return the same thing.  So the only safe approach we can take here is to assure that everything happening from cache-check to cache-put is protected by a critical section.

I worked with an engineer who had achieved lockless multithreading on a messaging server he'd built, so it is possible, but required him to write all the low level functions himself, even the memory management.

Reply:
Thanks for the explanation!